src/room007/models/word_tag_predictor.py

```diff
@@ -1,100 +1,130 @@
 # vim: set fileencoding=utf-8 :
 
-import itertools
-from sklearn.feature_extraction.text import TfidfVectorizer
-from sklearn import linear_model
-
-import nltk
+from copy import deepcopy
+from itertools import chain
+import logging
 import string
 
-from sklearn.neighbors import KNeighborsClassifier
-from sklearn.tree import DecisionTreeClassifier
-from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
-from sklearn.naive_bayes import GaussianNB
+logger = logging.getLogger(__name__)
+
+import nltk
+from sklearn import linear_model
 from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
+from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
+from sklearn.feature_extraction.text import TfidfVectorizer
 from sklearn.linear_model import LogisticRegression
+from sklearn.naive_bayes import GaussianNB
+from sklearn.neighbors import KNeighborsClassifier
+from sklearn.tree import DecisionTreeClassifier
 
 from room007.models import model
+from room007.util import time_function
+
 
+stop_words = (set(nltk.corpus.stopwords.words('english'))
+              .union(string.printable))
 
-stop_words = nltk.corpus.stopwords.words('english') + [x for x in string.printable]
+
+def identity(x):
+    return x
+
+
+def tokenize(sentence):
+    return [token for token in sentence.split() if token not in stop_words]
 
 
 class Features(object):
     def __init__(self, functional_test=False, changes=False):
         self.functional_test = functional_test
         self.tf_idf_vectorizer = None
         self.changes = changes
+        self._changes_int = 1 if changes else 0
+        self._fit_data_id = None
 
     def fit(self, train_data):
-        self._train_tf_idf_vectorizer(train_data)
+        if self._fit_data_id != id(train_data):
+            data_cp = deepcopy(train_data)
```

Comment:
Replying to the second comment: yes, but you skip the important part, which is the definition of the train dataset for cross-validation:

        train_dataset = pd.concat(frame for name, frame in frames.items()
                                  if name != eval_name)

Without creating a copy of the train dataset here in the `Features` class (assuming the derived columns have not been added earlier), adding new columns with derived features to the data frame modifies it in place, which causes the new train set in the second iteration of cross-validation to contain NaN's in that column for the sub-frame that was not part of the train set in the first iteration, causing bad things to happen.

I have implemented a better solution on my current branch, edit_scripts, which is a continuation of this one. I thought this MR would come in a good time when some coherent chunk of functionality is already added but we can as well wait until I am done with development on the edit_scripts branch and have a cleaner solution but presumably non-trivial merge conflicts.